pyomo/contrib/fbbt/expression_bounds_walker.py

```diff
@@ -0,0 +1,282 @@
+
+
+def _handle_sin(visitor, node, arg):
+    return sin(*arg)
+
+
+def _handle_cos(visitor, node, arg):
+    return cos(*arg)
+
+
+def _handle_tan(visitor, node, arg):
+    return tan(*arg)
+
+
+def _handle_asin(visitor, node, arg):
+    return asin(*arg, -pi / 2, pi / 2, visitor.feasibility_tol)
+
+
+def _handle_acos(visitor, node, arg):
+    return acos(*arg, 0, pi, visitor.feasibility_tol)
+
+
+def _handle_atan(visitor, node, arg):
+    return atan(*arg, -pi / 2, pi / 2)
+
+
+def _handle_sqrt(visitor, node, arg):
+    return power(*arg, 0.5, 0.5, feasibility_tol=visitor.feasibility_tol)
+
+
+def _handle_AbsExpression(visitor, node, arg):
+    return interval_abs(*arg)
+
+
+def _handle_UnaryFunctionExpression(visitor, node, arg):
+    return _unary_function_dispatcher[node.getname()](visitor, node, arg)
+
+
+def _handle_named_expression(visitor, node, arg):
+    visitor.leaf_bounds[node] = arg
+    return arg
+
+
+_unary_function_dispatcher = {
+    'exp': _handle_exp,
+    'log': _handle_log,
+    'log10': _handle_log10,
+    'sin': _handle_sin,
+    'cos': _handle_cos,
+    'tan': _handle_tan,
+    'asin': _handle_asin,
+    'acos': _handle_acos,
+    'atan': _handle_atan,
+    'sqrt': _handle_sqrt,
+}
+
+
+_operator_dispatcher = ExitNodeDispatcher(
+    {
+        ProductExpression: _handle_ProductExpression,
+        DivisionExpression: _handle_DivisionExpression,
+        PowExpression: _handle_PowExpression,
+        AbsExpression: _handle_AbsExpression,
+        SumExpression: _handle_SumExpression,
+        MonomialTermExpression: _handle_ProductExpression,
+        NegationExpression: _handle_NegationExpression,
+        UnaryFunctionExpression: _handle_UnaryFunctionExpression,
+        LinearExpression: _handle_SumExpression,
+        Expression: _handle_named_expression,
+    }
+)
+
+
+class ExpressionBoundsVisitor(StreamBasedExpressionVisitor):
+    """
+    Walker to calculate bounds on an expression, from leaf to root, with
+    caching of terminal node bounds (Vars and Expressions)
+
+    NOTE: If anything changes on the model (e.g., Var bounds, fixing, mutable
+    Param values, etc), then you need to either create a new instance of this
+    walker, or clear self.leaf_bounds!
+
+    Parameters
+    ----------
+    leaf_bounds: ComponentMap in which to cache bounds at leaves of the expression
+        tree
+    feasibility_tol: float, feasibility tolerance for interval arithmetic
+        calculations
+    use_fixed_var_values_as_bounds: bool, whether or not to use the values of
+        fixed Vars as the upper and lower bounds for those Vars or to instead
+        ignore fixed status and use the bounds. Set to 'True' if you do not
+        anticipate the fixed status of Variables to change for the duration that
+        the computed bounds should be valid.
+    """
+
+    def __init__(
+        self,
+        leaf_bounds=None,
+        feasibility_tol=1e-8,
+        use_fixed_var_values_as_bounds=False,
```

Comment:
I did False because it seems safer to me... Because then the bounds are right regardless of if variables get unfixed later. Basically this way we get positive confirmation from a user that they know things are potentially invalid if they unfix Vars.